### PR TITLE
Repository refactor

### PR DESCRIPTION
**Pull request by jubeira**
This script splits `do_everything` (`install.sh`) into smaller, maintainable scripts; the utilities are split into smaller files as well. I ran the script from scratch and it generated the output properly so far.

TODO
- [ ] Iron out `install` file below `build_cpp` command.
- [ ] Restore old `do_docker` with the current docker scripts.
- [ ] Sanity check: check that the right variables are being passed to the scripts as they run.
- [ ] Add a proper README file.
- [ ] Move exported variables to config file.

@meyerj let me know if this sounds reasonable to you at a high level, or if you strongly disagree with these changes; I still need to polish many details so I don't think a thorough review is worth it right now.

### CODE REVIEW

File: scripts/sources_setup/get_ros_stuff.sh [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+[ "$CMAKE_PREFIX_PATH" = "" ] && die 'could not find target basedir. Have you run build_catkin.sh and sourced setup.bash?'
+
+pushd $OUTPUT_PREFIX
+mkdir -p $OUTPUT_PREFIX/catkin_ws/src # && cd catkin_ws
+
+if [ -f $OUTPUT_PREFIX/catkin_ws/src/.rosinstall ]; then
+  pushd $OUTPUT_PREFIX/catkin_ws/src/
+  wstool merge $ROS_ANDROID_ROOT/ndk.rosinstall --merge-replace
+  wstool update
```
Comment (meyerj) [inline scripts/sources_setup/get_ros_stuff.sh:28]: add `-j$PARALLEL_JOBS` here, too

**Review by meyerj — COMMENTED** [review]
Looks good! I wonder whether the process could be simplified even further:

- Use `wstool` to get the system libraries - at least most of them. There is really no good reason to have an ugly shell script like [get_library.sh](https://github.com/Intermodalics/ros_android/pull/47/files#diff-54ca5be5ac5fd5b7c12ead198a9d6956) and [download_utils.sh](https://github.com/Intermodalics/ros_android/pull/47/files#diff-becd04a20c9ad3ba52ab5ef5045b8a96), which are basically reimplementing [wstool](http://wiki.ros.org/wstool). The few special cases could be handled in [get_system_libraries.sh](https://github.com/Intermodalics/ros_android/pull/47/files#diff-a1dea558d71ad4a8582b166247db6ac6) and `boost/build-android.sh` could be called from `build_library.sh`, too. I suggest to rename `ndk.rosinstall` to `ros.rosinstall` and add another file called `system_dependencies.rosinstall`. Ideally those two files should are the only place that mentions version numbers. In all other cases the version could be simply omitted or replaced by shell globs.

- The rosinstall files can be symlinked to their respective source-space `catkin_ws/src` or `libs`. No need to merge and no 'risk" of having dangling packages when switching branches and packages have been removed. However, the working copies would still need to be deleted manually after the removal of a package from the committed rosinstall file.

- The patches could be applied automatically simply by enumerating the files in `patches/*.patch`. No need to separate between system libraries and catkin packages or to list them explicitly in [get_patched_ros_workspace.sh](https://github.com/Intermodalics/ros_android/pull/47/files#diff-f20084e4851267140a86d7e9a0cf95c1). By the way, the patches applied there are not only applied to the ROS workspace as the name suggests, but also to system dependencies downloaded to and built in `libs/`.

- The build system to use for each library in [build_system_libraries.sh](https://github.com/Intermodalics/ros_android/pull/47/files#diff-7242c5b6cadb24637b2cabfc2cc9a1fe) could be detected automatically based on the existence of files like `CMakeLists.txt` or `configure`. For example, `build_library` could be a general command which forwards to `build_library_with_cmake` or `build_library_with_toolchain`. However, the order and the marker file to detect the presence of the package in the target-space still needs to be specified manually somewhere, so other than for patches the libraries still need to be listed explicitly here.

- Rename [build_cpp.sh](https://github.com/Intermodalics/ros_android/pull/47/files#diff-486689dfe70aeb0620ea48405f58a9b5) to `build_catkin_workspace.sh` or `build_ros_stuff.sh` (for consistency with `get_ros_stuff.sh`). In fact it builds not only ROS packages, but also opencv and other dependencies which happen to be catkinized. You could even catkinize all the other system libraries (by adding a `package.xml` and eventually a `CMakeLists.txt` file) and build everything using catkin, e.g. using [ExternalProject_Add()](https://cmake.org/cmake/help/latest/module/ExternalProject.html).

The goal would be that switching versions or adding packages can be done by simply adding them to one of the two rosinstall files and eventually adding or amending a patch. None of my suggestions is a must-have. The patch as-is is already an improvement.

**Comment by jubeira** [comment]
Thanks @meyerj, those are very thoughtful comments.
I agree with them; as a first stage I'd like to start using `wstool` to replace `get_libraries` script and applying all patches at once without explicitly listing them. I'd leave symlinking and merging `build_library*` scripts for another PR (I'd like to make sure that the existence of those files is enough to determine which one to use).

I'll be working on this. Let's try to merge the open PRs anyway; I can then rebase or merge and fix conflicts on this branch.

**Comment by jubeira** [comment]
Let's follow those proposals but in smaller PRs; this one has too many conflicts already .

**Pull request closed without merging**